test/core/derivatives/problem.py

```diff
@@ -95,6 +95,12 @@ def set_up(self):
         self.input_shape = self.make_input_shape()
         self.output_shape = self.make_output_shape()
 
+        if not self.module.training:
+            self.module.train(True)
+            for _ in range(3):
+                self.module(torch.randn_like(self.input))
+            self.module.train(False)
+
```

Comment:
What does this code do? Set up the statistics moving averages?

Reply:
Exactly.

Alternatively, we could set the running stats to some random values. Is this better?

Reply:
Advantage of the current approach is that the same pattern works for the extension problems.

Reply:
There are other modules, like `Dropout`, which we maybe want to test with `train(False)` in the future as they behave differently in evaluation mode. Maybe the `if` condition must be more specific to not run additional forward passes for those modules.

Reply:
Could also argue the other way around. It is safer to just run some forward passes to be sure that the derivatives stay correct. Thus, it is a trade-off between test execution time and safety.
I argue, that the cost of three additional forward passes is acceptable.

Addtional argument (same as above): For extension test problems, it would be very difficult to write an if-condition to test whether any child contains a BatchNorm module.

However, I could live with writing a specific if-condition in this case and leaving it general for the extensions.

Reply:
In my opinion, the criterion for the derivatives to work, no matter how many forward passes, is that they have the correct IO. This is done by PyTorch's `forward_hook` and should work reliably.

Then, the 'only' reason for additional forward passes is filling the moving averages. I currently favor the solution where this is done in the `SETTINGS` entry `'module_fn'`. Like this,
1. the other tests remain unaffected
2. no additional cost from forward passes
3. no need to write specific if conditions
but the code for specifying a setting becomes longer.

Feedback?

Reply:
Yes, I agree and will change it accordingly.

However, for future modules keep in mind the following:
I wrote the whole BatchNormDerivatives and tested it successfully. The next day, I realized that the running_stats are trivial and changed them and suddenly all tests failed.
I guess it is fine, because it was also a little bit stupid, and there are very few modules with such a behaviour.

Reply:
Just to be sure I understand the problem you had: The problem was that you didn't set the running stats to a value that differs from 0?

Reply:
Yes. For most modules the initialization is random. For BatchNorm it is: mean=0, var=1, weight=1, bias=0, which is not random at all.